### PR TITLE
Adds guide for writing custom quantization compatible layers

### PR DESCRIPTION
**Pull request by JyotinderSingh**

### CODE REVIEW

File: guides/ipynb/writing_quantization_compatible_layers.ipynb [hunk 1]
```diff
@@ -0,0 +1,798 @@
...
+    "\n",
+    "    quantized_kernel, scale = quantizers.abs_max_quantize(\n",
+    "        self._kernel, axis=0, to_numpy=True\n",
+    "    )\n",
+    "    scale = ops.squeeze(scale, axis=0)\n",
+    "\n",
+    "    del self._kernel\n",
+    "\n",
+    "    # Allocate INT8 variables. Discussed in the next section.\n",
+    "    self._int8_build(kernel_shape=self._kernel.shape)\n",
```
[the remark this answers is not in the document]
Comment (JyotinderSingh) [reply guides/ipynb/writing_quantization_compatible_layers.ipynb]: Fixed.

File: guides/ipynb/writing_quantization_compatible_layers.ipynb [hunk 2]
```diff
@@ -0,0 +1,798 @@
...
+   "cell_type": "markdown",
+   "metadata": {
+    "colab_type": "text"
+   },
+   "source": [
+    "### The `_int8_call(...)` method\n",
+    "\n",
+    "The `_int8_call(...)` method implements a minimal INT8 forward path. It uses the\n",
+    "quantized variables allocated in `_int8_build(...)` and de-scales the output\n",
+    "back to floating-point.\n",
```
[the remark this answers is not in the document]
Comment (JyotinderSingh) [reply guides/ipynb/writing_quantization_compatible_layers.ipynb:254]: Added.

File: guides/ipynb/writing_quantization_compatible_layers.ipynb [hunk 3]
```diff
@@ -0,0 +1,798 @@
...
+    "Here are concrete patterns you can reuse when making your own layers PTQ-friendly.\n",
+    "\n",
+    "- Build-time vs call-time responsibilities\n",
+    "  - In `build(...)`, if `self.quantization_mode` is set: allocate the quantized\n",
+    "    variables and skip allocating the float kernel to avoid duplicates.\n",
+    "- Record any metadata you need for the call path, e.g., for INT4:\n",
+    "  - The axis you packed along (e.g., `_int4_pack_axis`).\n",
+    "  - The original (unpacked) length on that axis (e.g., `_original_input_dim` or\n",
+    "    `_original_length_along_pack_axis`).\n",
+    "- In `call(...)`, compute with the quantized buffers and de-scale back to float\n",
```
[the remark this answers is not in the document]
Comment (JyotinderSingh) [reply guides/ipynb/writing_quantization_compatible_layers.ipynb]: Fixed.

File: guides/md/writing_quantization_compatible_layers.md [hunk 4]
```diff
@@ -0,0 +1,610 @@
...
+
+    quantized_kernel, scale = quantizers.abs_max_quantize(
+        self._kernel, axis=0, to_numpy=True
+    )
+    scale = ops.squeeze(scale, axis=0)
+
+    del self._kernel
+
+    # Allocate INT8 variables. Discussed in the next section.
+    self._int8_build(kernel_shape=self._kernel.shape)
```
[the remark this answers is not in the document]
Comment (JyotinderSingh) [reply guides/md/writing_quantization_compatible_layers.md]: Fixed.

File: guides/md/writing_quantization_compatible_layers.md [hunk 5]
```diff
@@ -0,0 +1,610 @@
...
+2. If you support INT4 quantization, implement a similar `_int4_build(...)`
+  method that allocates packed INT4 storage (often packed into INT8) plus
+  per-feature scales. The original unpacked dimensions and packing axis should
+  be recorded as instance variables for use in the call path.
+
+### The `_int8_call(...)` method
+
+The `_int8_call(...)` method implements a minimal INT8 forward path. It uses the
+quantized variables allocated in `_int8_build(...)` and de-scales the output
+back to floating-point.
```
[the remark this answers is not in the document]
Comment (JyotinderSingh) [reply guides/md/writing_quantization_compatible_layers.md:180]: Fixed.

File: guides/md/writing_quantization_compatible_layers.md [hunk 6]
```diff
@@ -0,0 +1,610 @@
...
+Here are concrete patterns you can reuse when making your own layers PTQ-friendly.
+
+- Build-time vs call-time responsibilities
+  - In `build(...)`, if `self.quantization_mode` is set: allocate the quantized
+    variables and skip allocating the float kernel to avoid duplicates.
+- Record any metadata you need for the call path, e.g., for INT4:
+  - The axis you packed along (e.g., `_int4_pack_axis`).
+  - The original (unpacked) length on that axis (e.g., `_original_input_dim` or
+    `_original_length_along_pack_axis`).
+- In `call(...)`, compute with the quantized buffers and de-scale back to float
```
[the remark this answers is not in the document]
Comment (JyotinderSingh) [reply guides/md/writing_quantization_compatible_layers.md]: Fixed.

File: guides/writing_quantization_compatible_layers.py [hunk 7]
```diff
@@ -0,0 +1,581 @@
...
+
+    quantized_kernel, scale = quantizers.abs_max_quantize(
+        self._kernel, axis=0, to_numpy=True
+    )
+    scale = ops.squeeze(scale, axis=0)
+
+    del self._kernel
+
+    # Allocate INT8 variables. Discussed in the next section.
+    self._int8_build(kernel_shape=self._kernel.shape)
```
[the remark this answers is not in the document]
Comment (JyotinderSingh) [reply guides/writing_quantization_compatible_layers.py]: Fixed.

File: guides/writing_quantization_compatible_layers.py [hunk 8]
```diff
@@ -0,0 +1,581 @@
...
+  per-feature scales. The original unpacked dimensions and packing axis should
+  be recorded as instance variables for use in the call path.
+"""
+
+"""
+### The `_int8_call(...)` method
+
+The `_int8_call(...)` method implements a minimal INT8 forward path. It uses the
+quantized variables allocated in `_int8_build(...)` and de-scales the output
+back to floating-point.
```
[the remark this answers is not in the document]
Comment (JyotinderSingh) [reply guides/writing_quantization_compatible_layers.py:179]: Fixed.

File: guides/writing_quantization_compatible_layers.py [hunk 9]
```diff
@@ -0,0 +1,581 @@
...
+Here are concrete patterns you can reuse when making your own layers PTQ-friendly.
+
+- Build-time vs call-time responsibilities
+  - In `build(...)`, if `self.quantization_mode` is set: allocate the quantized
+    variables and skip allocating the float kernel to avoid duplicates.
+- Record any metadata you need for the call path, e.g., for INT4:
+  - The axis you packed along (e.g., `_int4_pack_axis`).
+  - The original (unpacked) length on that axis (e.g., `_original_input_dim` or
+    `_original_length_along_pack_axis`).
+- In `call(...)`, compute with the quantized buffers and de-scale back to float
```
[the remark this answers is not in the document]
Comment (JyotinderSingh) [reply guides/writing_quantization_compatible_layers.py]: Fixed.

File: guides/md/writing_quantization_compatible_layers.md [hunk 10]
```diff
@@ -0,0 +1,612 @@
...
+
+```python
+
+def quantize(self, mode, **kwargs):
+    if mode != "int8":
+        raise NotImplementedError(f"Unsupported quantization mode: {mode}")
+
+    quantized_kernel, scale = quantizers.abs_max_quantize(
+        self._kernel, axis=0, to_numpy=True
+    )
```
Comment (hertschuh) [inline guides/md/writing_quantization_compatible_layers.md:99]: Can you explicitly pass `dtype="int8"` here? It's the default, but it makes you wonder how you would change this code to do "int4".

File: guides/md/writing_quantization_compatible_layers.md [hunk 11]
```diff
@@ -0,0 +1,612 @@
...
+  inform the setter that quantized variables are initialized.
+
+### The `_int8_build(...)` method
+
+This `int8_build(...)` method is called from `quantize(...)` to initialize the
+INT8 variables. It should allocate:
+
+- `self._kernel` as an INT8 vector of shape `(input_dim,)` (the same shape as
+  the original full-precision kernel).
+- `self.scale` as the scalar quantization scale in FP32.
```
Comment (hertschuh) [inline guides/md/writing_quantization_compatible_layers.md]: Because you don't pass the dtype when creating the `scale`, it inherits the dtype from the current dtype policy (before quantization), which is not necessarily FP32. I think you should mention that and then say it's FP32 in this example.

File: guides/md/writing_quantization_compatible_layers.md [hunk 12]
```diff
@@ -0,0 +1,612 @@
...
+        initializer="ones",
+        trainable=False,
+    )
+
+```
+
+#### Note
+
+1. The INT8 variables should be `trainable=False` since PTQ does not involve
+  further training.
```
Comment (hertschuh) [inline guides/md/writing_quantization_compatible_layers.md]: What happens if you make them trainable and do `model.fit()` do you get an error?

Doing `trainable=False` is a bit of a double edge sword. If you do train, what will happen is that it will only train non-quantized layers, which could be what you want, but maybe not. Maybe an error is better so that you know you can't train.

File: guides/md/writing_quantization_compatible_layers.md [hunk 13]
```diff
@@ -0,0 +1,612 @@
...
+2. If you support INT4 quantization, implement a similar `_int4_build(...)`
+  method that allocates packed INT4 storage (often packed into INT8) plus
+  per-feature scales. The original unpacked dimensions and packing axis should
+  be recorded as instance variables for use in the call path.
```
Comment (hertschuh) [inline guides/md/writing_quantization_compatible_layers.md]: I wouldn't know how to do that. Maybe link to the implementation in `dense.py` here too, like you did below?

File: guides/md/writing_quantization_compatible_layers.md [hunk 14]
```diff
@@ -0,0 +1,612 @@
...
+
+1. Elementwise multiply using the quantized weight.
+2. De-scale back to float by dividing with the `scale`.
+
+
+```python
+
+def _int8_call(self, inputs, training=None):
+    x = ops.multiply(inputs, self._kernel)
+    x = ops.cast(x, self.compute_dtype)
```
Comment (hertschuh) [inline guides/md/writing_quantization_compatible_layers.md]: The `multiply` above will promote to the inputs `dtype`, right?

And `self.scale` is of `compute_dtype`, right? Do you need the cast?

File: guides/md/writing_quantization_compatible_layers.md [hunk 15]
```diff
@@ -0,0 +1,612 @@
...
+    x = ops.divide(x, self.scale)
+    return x
+
+```
+
+---
+## Complete `SimpleScale` class with hooks
```
Comment (hertschuh) [inline guides/md/writing_quantization_compatible_layers.md:201]: It's unfortunate that you have to replicate the whole code 3 times.

File: guides/md/writing_quantization_compatible_layers.md [hunk 16]
```diff
@@ -0,0 +1,612 @@
...
+
+    def call(self, inputs, training=None):
+        return ops.multiply(inputs, self._kernel)
+
+    def quantize(self, mode, **kwargs):
+        if mode != "int8":
+            raise NotImplementedError(f"Unsupported quantization mode: {mode}")
+
+        quantized_kernel, scale = quantizers.abs_max_quantize(
+            self._kernel, axis=0, to_numpy=True
```
Comment (hertschuh) [inline guides/md/writing_quantization_compatible_layers.md]: Add `dtype="int8"` here too.

File: guides/md/writing_quantization_compatible_layers.md [hunk 17]
```diff
@@ -0,0 +1,612 @@
...
+## Extending to INT4
+
+If you want to support INT4 quantization, add:
+
+- `_int4_build(...)`: allocate a packed 4-bit storage (often packed into int8) plus per-feature scales
+- `_int4_call(...)`: unpack at runtime and follow the same de-scale pattern
+- `quantize("int4")`: quantize weights with `value_range=(-8, 7)`, then pack to int4 storage
+
+See the
+[Dense](https://github.com/keras-team/keras/blob/3c3d6adc08db627d89b5ad5e7f9b0ba3e88f2641/keras/src/layers/core/dense.py)
```
Comment (hertschuh) [inline guides/md/writing_quantization_compatible_layers.md]: Because this link is for a specific version of `dense.py` you can safely put a line number anchor if you want to point to `_int4_call(...)` for instance.

File: guides/md/writing_quantization_compatible_layers.md [hunk 18]
```diff
@@ -0,0 +1,612 @@
...
+
+    def call(self, inputs, training=None):
+        return ops.multiply(inputs, self._kernel)
+
+    def quantize(self, mode, **kwargs):
+        if mode != "int8":
+            raise NotImplementedError(f"Unsupported quantization mode: {mode}")
+
+        quantized_kernel, scale = quantizers.abs_max_quantize(
+            self._kernel, axis=0, to_numpy=True
```
Comment (hertschuh) [inline guides/md/writing_quantization_compatible_layers.md]: `dtype="int8"`

File: guides/md/writing_quantization_compatible_layers.md [hunk 19]
```diff
@@ -0,0 +1,612 @@
...
+            self.scale.assign(store[str(idx)])
+        else:
+            raise ValueError(f"Unsupported quantization mode for load: {mode}")
+
+```
+
+#### Note
+
+The `@keras.saving.register_keras_serializable()` decorator is needed to
+  register the class for serialization.
```
Comment (hertschuh) [inline guides/md/writing_quantization_compatible_layers.md]: nitpick: remove leading spaces.

File: guides/md/writing_quantization_compatible_layers.md [hunk 20]
```diff
@@ -0,0 +1,612 @@
...
+    dtype policy.
+  - Only change policy if the current policy has `quantization_mode is None` to
+    avoid an infinite loop.
+
+- Serialization contract
+    - Provide a fixed-order logic for variable serialization so save/load is
+      deterministic.
+    - Write variables in a fixed order per mode (e.g., None: [kernel, bias],
+      `"int8"`: [kernel, bias, kernel_scale], `"int4"`:
+      [kernel, bias, kernel_scale]).
```
Comment (hertschuh) [inline guides/md/writing_quantization_compatible_layers.md]: nitpick: this one is not indented like the other ones.

File: guides/md/writing_quantization_compatible_layers.md [hunk 21]
```diff
@@ -0,0 +1,612 @@
...
+to INT8 or INT4, and saved/loaded with quantization metadata.
+
+A quantization-compatible layer needs to implement a few hooks, so that it can:
+
+- Switch its variables to quantized representations.
+- Use a quantization-aware forward path at inference.
+- Save and load quantization metadata with the model.
+
+In this guide, we'll implement a simple layer that supports INT8 PTQ. The same
+patterns generalize to INT4 quantization and FP8 mixed-precision training.
```
Comment (hertschuh) [inline guides/md/writing_quantization_compatible_layers.md:29]: Later in the guide, you explain how to do INT4, and in particular you mention `_int4_build` the `_int4_call`.

However, you never talk about FP8 anymore, and in particular, the names `_float8_build` and `_float8_call` are not mentioned. I think you should at least mention those 2 method names. Initially I thought a table format might be good to show FP8 / INT8 / INT4 but I'll leave it up to you.

File: guides/md/writing_quantization_compatible_layers.md (hunk 10 above)
Comment (JyotinderSingh) [reply guides/md/writing_quantization_compatible_layers.md:99]: Done.

File: guides/md/writing_quantization_compatible_layers.md (hunk 11 above)
Comment (JyotinderSingh) [reply guides/md/writing_quantization_compatible_layers.md]: Modified the description

File: guides/md/writing_quantization_compatible_layers.md (hunk 13 above)
Comment (JyotinderSingh) [reply guides/md/writing_quantization_compatible_layers.md]: Added.

File: guides/md/writing_quantization_compatible_layers.md (hunk 14 above)
Comment (JyotinderSingh) [reply guides/md/writing_quantization_compatible_layers.md]: You're correct, we can skip the cast.

File: guides/md/writing_quantization_compatible_layers.md (hunk 15 above)
Comment (JyotinderSingh) [reply guides/md/writing_quantization_compatible_layers.md:201]: I've replicated it for clarity so that the readers always have a complete reference implementation to compare against. Open to other suggestions.

File: guides/md/writing_quantization_compatible_layers.md (hunk 16 above)
Comment (JyotinderSingh) [reply guides/md/writing_quantization_compatible_layers.md]: Done.

File: guides/md/writing_quantization_compatible_layers.md (hunk 17 above)
Comment (JyotinderSingh) [reply guides/md/writing_quantization_compatible_layers.md]: Good point, added.

File: guides/md/writing_quantization_compatible_layers.md (hunk 18 above)
Comment (JyotinderSingh) [reply guides/md/writing_quantization_compatible_layers.md]: Done.

File: guides/md/writing_quantization_compatible_layers.md (hunk 19 above)
Comment (JyotinderSingh) [reply guides/md/writing_quantization_compatible_layers.md]: Added.

File: guides/md/writing_quantization_compatible_layers.md (hunk 20 above)
Comment (JyotinderSingh) [reply guides/md/writing_quantization_compatible_layers.md]: Fixed.

File: guides/md/writing_quantization_compatible_layers.md (hunk 21 above)
Comment (JyotinderSingh) [reply guides/md/writing_quantization_compatible_layers.md:29]: FP8 is a mixed precision training technique that happens to do a quantization step in the middle. It includes more sophisticated variable history tracking and fake-quant injection which might be irrelevant to the particular audience of this guide.

File: guides/md/writing_quantization_compatible_layers.md (hunk 12 above)
Comment (JyotinderSingh) [reply guides/md/writing_quantization_compatible_layers.md]: Returning an error might not be the right thing to do, since that might break the fine-tuning (for non-quantized layers) and QLoRA (for quantized layers) flow. I've modified the comment to be clearer that it's the quantized variables that should no longer be modified, and not the rest of the model.

Thanks for catching this.

File: guides/md/writing_quantization_compatible_layers.md [hunk 22]
```diff
@@ -0,0 +1,613 @@
...
+  inform the setter that quantized variables are initialized.
+
+### The `_int8_build(...)` method
+
+This `int8_build(...)` method is called from `quantize(...)` to initialize the
+INT8 variables. It should allocate:
+
+- `self._kernel` as an INT8 vector of shape `(input_dim,)` (the same shape as
+  the original full-precision kernel).
+- `self.scale` as the scalar quantization scale in the layer's compute dtype,
```
Comment (hertschuh) [inline guides/md/writing_quantization_compatible_layers.md]: Technically, it's the "variable dtype" of the dtype policy, which is in theory allowed to be different from the "compute dtype".

**Review by hertschuh — APPROVED** [review]
Thanks for the guide!

File: guides/md/writing_quantization_compatible_layers.md (hunk 22 above)
Comment (JyotinderSingh) [reply guides/md/writing_quantization_compatible_layers.md]: Fixed.
